File: python/src/threads/send_thread.py
import datetime
from PyQt5 import QtCore
from PyQt5.QtCore import QObject, pyqtProperty, pyqtSignal, pyqtSlot

from src.app.canbus import CanBus


class SendThread(QObject):
    sendDone = pyqtSignal()
    sendError = pyqtSignal()

    def __init__(self, parent=None):
        super(SendThread, self).__init__(parent)

        self.can_bus = CanBus()

    @pyqtSlot(str, str, str)
    def send(self, can_id, dlc, hex_string):
        data_len = len(hex_string) // 2
        if data_len != dlc:
            self.sendError.emit()

        # hex string to int
        # -> With the 0x prefix, Python can distinguish hex and decimal automatically
        can_id = int(can_id, 0)
        dlc = int(dlc)
        byte_data = bytes.fromhex(hex_string)
        self.can_bus.transmit(can_id, dlc, byte_data)

        self.sendDone.emit()

File: cpp/src/threads/send_thread.cpp
#include "send_thread.h"

SendThread::SendThread()
{
  can_bus_ = new CanBus();
  can_bus_->open();
}

void SendThread::send(const QString &can_id, const QString &dlc,
                      const QString &data)
{
  can_id_ = can_id.toUInt(nullptr, 16);
  dlc_ = dlc.toUInt();
  int data_length = data.length();
  if (data_length / 2 != dlc_) {
    emit sendErrMsg("Data length wrong, please check it");
    return;
  }

  can_data_.reserve(dlc_);
  for (int i = 0; i < data_length; i += 2) {
    can_data_.push_back(data.mid(i, 2).toUInt(nullptr, 16));
  }

  msg_.can_id = can_id_;
  msg_.can_dlc = dlc_;
  msg_.data = can_data_;
  can_bus_->send(msg_);
  emit sendDone();
}
